### PR TITLE
fix: useModal return type error  #833

### PR DESCRIPTION
**Pull request by DaiQiangReal**
[中文模板 / Chinese Template](https://github.com/DouyinFE/semi-design/blob/main/.github/PULL_REQUEST_TEMPLATE.zh-CN.md)

- [x] I have read and followed [Pull Request Guidelines](https://github.com/DouyinFE/semi-design/blob/main/CONTRIBUTING-en-US.md#pull-request-guidelines) of the contributing guide.

### What kind of change does this PR introduce? (check at least one)

 - [x] Bugfix
 - [ ] Feature
 - [ ] Code style update
 - [ ] Refactor
 - [ ] Test Case
 - [ ] TypeScript definition update
 - [ ] Document improve
 - [ ] CI/CD improve
 - [ ] Branch sync
 - [ ] Other, please describe:

### PR description

Fixes #

### Changelog
🇨🇳 Chinese
- Fix: 修复 useModal typescript 返回类型不严谨的问题

---

🇺🇸 English
- Fix: Fix the problem that the return type of useModal typescript is not strict

### Checklist
- [x] Test or no need
- [ ] Document or no need
- [ ] Changelog or no need

### Other
- [ ] Skip Changelog

### CODE REVIEW

File: packages/semi-ui/modal/useModal/index.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ function usePatchElement(): ([ReactNode[], (element: ReactNode) => () => void])
     return [elements, patchElement];
 }
 
-export default function useModal() {
+export default function useModal(): [{ info: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void }, success: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void }, error: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void }, warning: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void }, confirm: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void } }, ReactNode] {
```
Comment (pointhalo) [inline packages/semi-ui/modal/useModal/index.tsx:22]: 这么长不换个行吗，看起来太费劲了

Comment (DaiQiangReal) [reply packages/semi-ui/modal/useModal/index.tsx]: 更新了

File: packages/semi-ui/modal/useModal/index.tsx [hunk 2]
```diff
@@ -19,7 +19,13 @@ function usePatchElement(): ([ReactNode[], (element: ReactNode) => () => void])
     return [elements, patchElement];
 }
 
-export default function useModal() {
+export default function useModal(): [{
+    info: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void },
+    success: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void },
+    error: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void },
+    warning: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void },
+    confirm: (config: ModalReactProps) => { destroy: () => void, update: (newConfig: ConfirmProps) => void }
```
Comment (tianenpang) [inline packages/semi-ui/modal/useModal/index.tsx:27]: 一个相关的问题：`ConfirmProps` 中的 `type` 由于不是可选属性，在 `modal.update` 时会提示 `Property 'type' is missing in type`，可能需要再处理下 🤔

model update-type-issue

**Review by tianenpang — CHANGES_REQUESTED** [review]

File: packages/semi-ui/modal/useModal/index.tsx [hunk 3]
```diff
@@ -19,7 +19,13 @@ function usePatchElement(): ([ReactNode[], (element: ReactNode) => () => void])
     return [elements, patchElement];
 }
 
-export default function useModal() {
+export default function useModal(): [{
```
[the remark this answers is not in the document]
Comment (DaiQiangReal) [reply packages/semi-ui/modal/useModal/index.tsx:24]: resolve

File: packages/semi-ui/modal/useModal/index.tsx (hunk 2 above)
Comment (DaiQiangReal) [reply packages/semi-ui/modal/useModal/index.tsx]: > 一个相关的问题：`ConfirmProps` 中的 `type` 由于不是可选属性，在 `modal.update` 时会提示 `Property 'type' is missing in type`，可能需要再处理下
> 
> model update-type-issue
> 
> 目前在非“确认模态对话框”时配置类型依然可以改变其类型
> 
> modal-type
> 
> 建议按以下类型约束会不会好些 🤔 `info, success, error, warning` 的 `update` 参数类型为 `ModalReactProps`
> 
> `confirm` 的 `update` 参数类型为 `ConfirmProps` 且 `type` 可选

可以再看下，刚加了泛型判断

**Pull request merged**